### ISSUE TITLE
Help with pycaffe

### ISSUE THREAD

**Issue by cmcuza** [alterzero/DBPN-caffe#1]
Hi, thanks for sharing the code.

I am trying to test your pre trained model in pycaffe however I must be doing something wrong because the results is not as expected. Here is the code and the result.

scale = 4.0
test_img_path = "butterfly_GT.bmp"
net = caffe.Net("model/DBPN_mat_4x.prototxt", "model/DBPN_4x.caffemodel", caffe.TEST)

img = cv2.imread(test_img_path, cv2.IMREAD_COLOR)
img = modcrop(img, scale)
lr_img = cv2.resize(img, None, fx=1/scale, fy=1/scale, interpolation=cv2.INTER_CUBIC) / 255.0
lr_img = lr_img[np.newaxis, :, :, :]
b, w, h, c = lr_img.shape
lr_img = np.reshape(lr_img, [b, c, w, h])

net.blobs['data'].reshape(*lr_img.shape)
net.blobs['data'].data[...] = lr_img
out = net.forward()
pred_img = np.asarray(out['conv_final'])
pred_img = pred_img.squeeze()
pred_img = np.reshape(pred_img, [img.shape[0], img.shape[1], img.shape[2]])

![result](https://user-images.githubusercontent.com/38991005/41621503-d74b96fe-73da-11e8-97cc-b9cb5cf21255.jpg)

**Comment by alterzero** [comment]
Hi,
Sorry for my very late reply. Can you try this code?

##################
import numpy as np
from PIL import Image
import caffe
import sys
import scipy.io
import scipy.misc
import timeit

caffe.set_device(0)
caffe.set_mode_gpu()

inputsize=64
upscale=4

net = caffe.Net("model/DBPN_mat_4x.prototxt", "model/DBPN_4x.caffemodel",caffe.TEST)

transformer = caffe.io.Transformer({'data': net.blobs['data'].data.shape})
transformer.set_transpose('data', (2,0,1))
net.blobs['data'].reshape(1,3,inputsize,inputsize)

im = caffe.io.load_image('baby_GT_input.png', color=True)
net.blobs['data'].data[...] = transformer.preprocess('data', im)

start = timeit.default_timer()
out = net.forward()
stop = timeit.default_timer()
print stop - start 

result=net.blobs['conv_final'].data

max_val, min_val = np.max(result[0]), np.min(result[0])

flatten_img = ((result.transpose((0,2,3,1)))[:] - min_val) / (max_val-min_val)

scipy.misc.imsave('result.png', flatten_img.reshape(1,1,inputsize*upscale,inputsize*upscale,3).swapaxes(1,2).reshape(inputsize*upscale,inputsize*upscale, 3))

**Comment by cmcuza** [comment]
Hi,

Thank you for your reply, it works perfect now !!!!!

**Comment by ynieWXJB** [comment]
Hi, 

I just copied the scripts you provided and did not change anything of them.

But I got an error as follows:
____________________________________________
File "firsttrial.py", line 17, in <module>
    net.blobs['data'].data[...] = transformer.preprocess('data', im)
ValueError: could not broadcast input array from shape (3,57,86) into shape (1,3,64,64)

_______________________________________________


I don't know why you can run it without problem. Did you change some thing in the scripts?


Many thanks

**Comment by cmcuza** [comment]
Hi, you are right I had to change dimension in the prototxt, change 57 and 86 to 64.

**Comment by ynieWXJB** [comment]
Thanks for your message.

After I changed these two parameters, I can run the code without errors. 

Did you get the same output as that of pytorch version?

I found that the result of pycaffe is worse than that of pytorch.

![dbpn_caffe_x4 _result](https://user-images.githubusercontent.com/40223583/43706276-76944576-9997-11e8-9c69-95b80f04eea2.png)
![dbpn_pytorch_x4_result](https://user-images.githubusercontent.com/40223583/43706277-76cfa378-9997-11e8-814c-4906f31bdf5b.png)
